File: sdks/rust/src/sandbox.rs
//! A sandbox, and the inbox that is the only way into it
//! (`DECISIONS.md#ports-are-the-hosts`, `drivers`).
//!
//! `Sandbox` is a HANDLE, not an instance. The instance lives wherever its
//! driver put it, and the handle is an id plus a way to reach it -- which is
//! what lets the driver decide the thread, and eventually the threads.
//!
//! A call does not run the program. It encodes a request, puts it in the
//! inbox, and wakes the driver; the driver decides when a runnable sandbox
//! runs, and the answer arrives on the `Pending` the call handed back.

use crate::driver::{Driver, Inline};
use crate::value::Value;
use crate::{Error, Result};
use flint_rt::native::Program;
use std::collections::VecDeque;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{mpsc, Arc, Mutex};
use std::time::Duration;

/// One request waiting to be dispatched.
struct Request {
    name: String,
    args: Vec<Value>,
    reply: mpsc::Sender<Result<Value>>,
}

/// The sandbox itself: its program, its inbox, and whether a dispatch is
/// already on its way.
///
/// Shared by every thread that may advance it, which is why the program is
/// behind a lock. **That lock is the whole of what stands between this and
/// several threads running one heap in parallel.** Removing it needs a
/// collector with per-thread allocation buffers and safepoints
/// (`DECISIONS.md#drivers`); it does not need a different interface, which is
/// why the interface is already written for K > 1.
pub struct Core {
    /// One executor per extra driver thread, each on the SAME heap.
    ///
    /// **Declared before `program` on purpose.** Struct fields drop in
    /// declaration order, and every executor borrows the heap the program
    /// owns, so dropping the program first would leave them pointing at freed
    /// memory. The order of these two lines is load-bearing.
    ///
    /// Handed out one-per-thread and never shared, which is the invariant that
    /// makes them safe: an executor is one thread's value stack, frames and
    /// gas, and the heap underneath is what the safepoint and the allocation
    /// lock coordinate (`DECISIONS.md#drivers`).
    ///
    /// `Mutex<Option<Box<Rt>>>` per slot rather than one lock over all of them,
    /// so two threads claiming different executors never meet.
    executors: Vec<Mutex<Option<Box<flint_rt::rt::Rt>>>>,
    program: Mutex<Program>,
    inbox: Mutex<VecDeque<Request>>,
    /// True from the moment a wake is accepted until the dispatch that owns it
    /// finds nothing left to do. THE coalescing flag: it answers true exactly
    /// once between dispatches, so a burst of arrivals costs one crossing of
    /// the sandbox boundary rather than one per arrival.
    scheduled: AtomicBool,
    /// How many times a dispatch has crossed into this sandbox, and how many
    /// requests those dispatches carried.
    ///
    /// Counted so the coalescing is MEASURABLE. "A burst of arrivals costs one
    /// crossing" is the claim the debouncing exists to make, and a claim with
    /// no number behind it is a hope.
    dispatches: std::sync::atomic::AtomicU64,
    served: std::sync::atomic::AtomicU64,
    /// Dispatches that ran on a SECONDARY executor rather than the program's
    /// own. Counted because "the pool is parallel" is otherwise unfalsifiable
    /// from outside: falling back to the primary passes every correctness test
    /// while running exactly as serially as before.
    on_executors: std::sync::atomic::AtomicU64,
}

impl Core {
    /// Claim the right to queue this sandbox. True exactly once per dispatch.
    pub(crate) fn schedule(&self) -> bool {
        !self.scheduled.swap(true, Ordering::AcqRel)
    }

    /// Run whatever is in the inbox, then whatever arrived while doing so.
    ///
    /// Called by a driver, never by a caller. Several driver threads may be in
    /// here at once for the same sandbox: they take separate batches and the
    /// program lock serialises the execution.
    pub fn advance(&self) {
        loop {
            let batch: Vec<Request> = {
                let mut inbox = self.inbox.lock().unwrap();
                inbox.drain(..).collect()
            };

            if batch.is_empty() {
                // Clearing and re-checking, in that order, and both. A request
                // can land between the drain above and this store; clearing
                // without looking again would leave it queued behind a flag
                // that says a dispatch is already coming, and nothing would
                // ever come. That is the lost wake-up, and it presents as a
                // sandbox that stops answering under load and not otherwise.
                self.scheduled.store(false, Ordering::Release);
                if self.inbox.lock().unwrap().is_empty() {
                    return;
                }
                if !self.schedule() {
                    // Somebody else took it; it is their batch now.
                    return;
                }
                continue;
            }

            self.dispatches.fetch_add(1, Ordering::Relaxed);
            self.served.fetch_add(batch.len() as u64, Ordering::Relaxed);

            // Take a spare executor if one is free. Several threads then run
            // guest code on ONE heap at the same time, which is the whole
            // point; the heap is kept consistent by the allocation lock and
            // the safepoint rather than by a lock around execution.
            //
            // Falling back to the primary when none is free is not a
            // compromise: with no spare executors -- an inline driver, or a
            // pool of one -- this is exactly what it did before.
            if let Some((slot, mut rt)) = self.claim_executor() {
                self.on_executors.fetch_add(1, Ordering::Relaxed);
                for request in batch {
                    let answer = dispatch_on(&mut rt, &request);
                    let _ = request.reply.send(answer);
                }
                *self.executors[slot].lock().unwrap() = Some(rt);
            } else {
                let mut program = self.program.lock().unwrap();
                for request in batch {
                    let answer = dispatch(&mut program, &request);
                    // A dropped receiver means the caller stopped waiting.
                    // That is allowed and is not an error: the work was still
                    // done, and its effect on the sandbox stands.
                    let _ = request.reply.send(answer);
                }
            }
        }
    }
}

impl Core {
    /// Take a free executor, or `None` if every one is busy.
    ///
    /// `try_lock` rather than `lock`: a busy slot means another driver thread
    /// is inside that executor, and waiting for it would serialise exactly the
    /// thing this exists to parallelise.
    fn claim_executor(&self) -> Option<(usize, Box<flint_rt::rt::Rt>)> {
        for (i, slot) in self.executors.iter().enumerate() {
            if let Ok(mut held) = slot.try_lock() {
                if let Some(rt) = held.take() {
                    return Some((i, rt));
                }
            }
        }
        None
    }
}

fn encode_call(request: &Request) -> Vec<u8> {
    let mut call = vec![Value::str(&request.name)];
    call.extend_from_slice(&request.args);
    Value::Vector(call).encode()
}

fn answer(out: core::result::Result<Vec<u8>, String>) -> Result<Value> {
    match out {
        Ok(bytes) => Value::decode(&bytes).map_err(Error::Encoding),
        Err(e) => {
            let (kind, message) = match e.split_once(": ") {
                Some((k, m)) => (k.to_string(), m.to_string()),
                None => ("Error".to_string(), e),
            };
            Err(Error::Call { kind, message })
        }
    }
}

fn dispatch(program: &mut Program, request: &Request) -> Result<Value> {
    answer(program.call(&encode_call(request)))
}

/// The same call, on a secondary executor. One protocol, two entry points --
/// `flint_rt::native::call_on` is the protocol, and both go through it.
fn dispatch_on(rt: &mut flint_rt::rt::Rt, request: &Request) -> Result<Value> {
    answer(flint_rt::native::call_on(rt, &encode_call(request)))
}

/// An answer that has not arrived yet.
///
/// Deliberately not a `std::future::Future`: this crate has no async runtime
/// and should not require one. It is the smallest thing every SDK can mirror
/// -- a promise in JavaScript, a `Task` in C#, a `CompletableFuture` in Java, a
/// completion callback in C -- and a caller who wants a real future can wrap it
/// in three lines.
#[must_use = "a call that is never awaited still runs; drop it deliberately"]
pub struct Pending<T> {
    rx: mpsc::Receiver<Result<T>>,
}

impl<T> Pending<T> {
    /// Block until the answer arrives.
    ///
    /// Never call this from inside a sandbox. A call made from guest code that
    /// is occupying a driver thread, waiting on a reply that needs a driver
    /// thread, is the oldest deadlock there is -- which is why from inside, a
    /// call is a port send and a park rather than this.
    pub fn wait(self) -> Result<T> {
        self.rx.recv().map_err(|_| Error::Call {
            kind: "Dropped".into(),
            message: "the sandbox went away before answering".into(),
        })?
    }

    /// Block for at most `timeout`. `None` means it has not arrived, not that
    /// it failed -- the request is still in flight and the answer is lost.
    pub fn wait_timeout(self, timeout: Duration) -> Option<Result<T>> {
        self.rx.recv_timeout(timeout).ok()
    }

    /// The answer if it is already here, without blocking.
    pub fn try_take(&self) -> Option<Result<T>> {
        self.rx.try_recv().ok()
    }
}

/// A running instance of an image, addressed through its driver.
///
/// Cheap to clone: a clone is another handle to the same sandbox, which is what
/// lets several threads hold it.
#[derive(Clone)]
pub struct Sandbox {
    core: Arc<Core>,
    driver: Arc<dyn Driver>,
}

impl Sandbox {
    /// From a compiled `.wasm` artifact, run inline.
    pub fn from_wasm(wasm: &[u8]) -> Result<Sandbox> {
        Sandbox::from_wasm_with(wasm, Arc::new(Inline))
    }

    pub fn from_wasm_with(wasm: &[u8], driver: Arc<dyn Driver>) -> Result<Sandbox> {
        let bytecode = flint_rt::native::bytecode_of(wasm).map_err(Error::Load)?;
        Sandbox::from_bytecode_with(&bytecode, driver)
    }

    /// From flint's own bytecode. An implementation detail, exposed because the
    /// CLI has one to hand and going back through wasm would be silly.
    pub fn from_bytecode(bytecode: &[u8]) -> Result<Sandbox> {
        Sandbox::from_bytecode_with(bytecode, Arc::new(Inline))
    }

    pub fn from_bytecode_with(bytecode: &[u8], driver: Arc<dyn Driver>) -> Result<Sandbox> {
        let mut program = Program::load(bytecode, 2_000_000_000).map_err(Error::Load)?;

        // One spare executor per EXTRA driver thread. The thread that gets
        // none falls back to the program's own, so a pool of N runs N-way and
        // an inline driver allocates nothing.
        //
        // Made here rather than lazily because an executor registers the
        // address of its root stack with the collector, and a slot handed out
        // and put back is one address for the sandbox's life.
        let mut executors = Vec::new();
        for _ in 1..driver.parallelism() {
            // SAFETY: every executor is dropped in `Core::drop`, before the
            // program that owns the heap they share.
            match unsafe { program.executor() } {
                Some(rt) => executors.push(Mutex::new(Some(rt))),
                None => break,
            }
        }

        Ok(Sandbox {
            core: Arc::new(Core {
                executors,
                program: Mutex::new(program),
                inbox: Mutex::new(VecDeque::new()),
                scheduled: AtomicBool::new(false),
                dispatches: std::sync::atomic::AtomicU64::new(0),
                served: std::sync::atomic::AtomicU64::new(0),
                on_executors: std::sync::atomic::AtomicU64::new(0),
            }),
            driver,
        })
    }

    /// How many threads may be inside this sandbox at once.
    ///
    /// Read it rather than assume it: a pool of 4 on a target that cannot
    /// honour it answers 1.
    pub fn parallelism(&self) -> usize {
        self.driver.parallelism()
    }

    /// Call a function by name with positional arguments.
    ///
    /// **This does not run the program.** It queues a request and wakes the
    /// driver; the answer arrives on the `Pending`. It has to be this way for a
    /// sandbox that may not run on this thread at all -- and it is asynchronous
    /// even under the inline driver, because a synchronous API cannot be made
    /// asynchronous later without breaking every caller.
    ///
    /// What an argument MEANS is the caller's business: there is no entry map
    /// here and no capability argument. Those are the CLI's convention
    /// (`DECISIONS.md#structured-ports`).
    pub fn call(&self, name: &str, args: &[Value]) -> Pending<Value> {
        let (tx, rx) = mpsc::channel();
        {
            let mut inbox = self.core.inbox.lock().unwrap();
            inbox.push_back(Request {
                name: name.to_string(),
                args: args.to_vec(),
                reply: tx,
            });
        }
        // Queued BEFORE the wake, always. A wake that arrives before the work
        // it announces can be answered by a dispatch that finds an empty inbox
        // and goes back to sleep.
        self.driver.wake(Arc::clone(&self.core));
        Pending { rx }
    }

    /// `call`, waited on. Convenience for a host thread with nothing else to
    /// do -- never reachable from inside a sandbox, where a call is a send and
    /// a park.
    pub fn call_blocking(&self, name: &str, args: &[Value]) -> Result<Value> {
        self.call(name, args).wait()
    }

    /// How many times a dispatch has crossed into this sandbox, against how
    /// many requests were served: `(dispatches, requests)`.
    ///
    /// The second divided by the first is what the debouncing bought. A host
    /// that sees them equal under load is getting no coalescing at all, which
    /// is worth being able to find out rather than guess.
    pub fn dispatch_counts(&self) -> (u64, u64) {
        (
            self.core.dispatches.load(Ordering::Relaxed),
            self.core.served.load(Ordering::Relaxed),
        )
    }

    /// How many dispatches ran on a secondary executor -- that is, genuinely
    /// alongside another thread inside the interpreter rather than behind the
    /// program lock.
    ///
    /// Readable because otherwise there is no way to tell a parallel pool from
    /// one that quietly fell back to serialising, and the two pass identical
    /// correctness tests.
    pub fn parallel_dispatches(&self) -> u64 {
        self.core.on_executors.load(Ordering::Relaxed)
    }

    /// PROJECT named opaque values into the sandbox, as the entry's second
    /// argument.
    ///
    /// This used to be `grant(name)`, backed by a table the RUNTIME kept -- and
    /// that made the runtime the arbiter of what a capability was. It is not,
    /// and now it does not know the word. What crosses is an ordinary opaque
    /// value (`DECISIONS.md#opaque-values`) carrying an id you chose; guest code cannot
    /// mint that id, so you recognise your own and nothing else. Whether it
    /// means a capability is entirely yours to decide, and a host that requires
    /// none passes nothing.

    /// A bound on WORK, in instructions. Deterministic
    /// (`DECISIONS.md#resource-limits`), so the same program stops at the same
    /// instruction on every machine.
    pub fn set_step_limit(&self, n: u64) {
        self.core.program.lock().unwrap().set_step_limit(n);
    }

    /// Instructions executed so far -- **only while a step limit is set**.
    ///
    /// This reads 0 on a sandbox with no limit, and that is not a bug to route
    /// around. Counting every instruction would put an increment and a compare
    /// in the interpreter's inner loop for every program, including the ones
    /// that never ask; instead the unbudgeted loop has no counter at all and
    /// the optimiser deletes the check (`DECISIONS.md#resource-limits`). So
    /// `set_step_limit` is what turns counting on.
    ///
    /// It is exact because the program lock makes it exact. When several
    /// threads really run one heap, gas becomes per-executor and summed -- a
    /// single shared counter would put an atomic on the hottest line in the
    /// interpreter -- and this will then be a snapshot true at a safepoint.
    pub fn gas(&self) -> u64 {
        self.core.program.lock().unwrap().steps()
    }
}

File: runtimes/jvm/src/com/flint/rt/Sandbox.java
package com.flint.rt;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/// THE THREE OPERATIONS, on the JVM (`DECISIONS.md#four-operations`).
///
/// `boot`, `loop`, `link`. The semantics are the record's and are identical on
/// every target; only the spelling is this one's.
///
/// ## `prop` WAS THE FOURTH AND IS GONE
///
/// The metadata is a class ATTRIBUTE on the generated `flint.Artifact`, named
/// `com.3sln.flint.meta` -- the same string the wasm custom section uses. All three
/// container formats carry metadata a reader gets at WITHOUT executing the
/// artifact, and a method call never can. So nothing here answers questions about
/// the artifact: a reader with the class file has the answers already, and one
/// without it was never going to get them from a method.
///
/// Reading it needs no JVM and no class loading -- JVMS 4.7 requires an
/// unrecognised attribute to be silently ignored, which is what makes a custom one
/// legal and inert, and also what makes the namespace load-bearing: a colliding
/// attribute name is not an error, it is wrong data read as right.
///
/// ## The artifact carries the PROGRAM, not the runtime
///
/// "Self-contained" means *contains all of the program's code*, NOT *runs
/// standalone*. This class, the collector and the builtins are the HOST's: an
/// artifact is one class file carrying bytecode and the three operations, and it
/// needs a JVM with flint's runtime on the classpath. Saying it the other way
/// round is the mistake the record warns about first.
///
/// ## Where this differs from wasm, and why that is allowed
///
/// A JVM sandbox can be handed a real OBJECT, so the bridge is one `Bridge` with
/// five methods over a `(port, bytes)` pair -- mirroring the CLR's `TryTake`,
/// `Put`, `Open`, `Answer`, `Closed`. wasm needs fourteen exports because most of
/// them exist only to move bytes across a boundary that cannot pass an array or a
/// callback: `flint_in_alloc` plus `flint_deliver` is `Put` inverted,
/// `flint_drain` plus `flint_events_ptr` plus five-`u32` record parsing is
/// `TryTake`, and `flint_grant`/`flint_answer`/`flint_continue` are `Open` and
/// `Answer` with the return value replaced by a token the host has to keep.
///
/// THE INBOUND DIRECTION IS A PULL, and that is the part worth noticing. The
/// sandbox asks the bridge for the next message rather than the host pushing one
/// in, so the host never needs a handle on the sandbox to deliver -- which is
/// what lets `boot` take one argument and still serve every port the host owns.
/// An earlier draft of this file pushed, and needed a fifth operation to attach
/// each further port to the sandbox first.
public final class Sandbox {

    // --- the bridge --------------------------------------------------------

    /// A message on a port. The `(port, bytes)` pair the five methods work over.
    ///
    /// BYTES, not text: a port's format may be binary (Transit over msgpack is),
    /// so decoding as UTF-8 here would replace whatever is not valid and corrupt
    /// the message.
    public record Msg(int port, byte[] bytes) {}

    /// The host's whole side of the bridge.
    ///
    /// ONE OBJECT FOR EVERY PORT, not one per port: the id travels in the
    /// arguments, which is what `(port, bytes)` means. The id is the HOST's, and
    /// that is the inversion the port design rests on
    /// (`DECISIONS.md#ports-are-the-hosts`) -- the sandbox does not mint an
    /// endpoint and offer it up, so an id means the same thing in every sandbox
    /// holding it.
    public abstract static class Bridge {
        /// The port that becomes this sandbox's SYSTEM port -- the one it can ask
        /// the host for things through. A sandbox given none runs logic and can
        /// ask for nothing, which is a coherent thing to be.
        public final int systemPort;
        public final String label;

        public Bridge(int systemPort, String label) {
            this.systemPort = systemPort;
            this.label = label;
        }

        /// HOST TO GUEST, pulled. The next message the host has for the sandbox,
        /// or null when it has none. `loop` calls this until it answers null.
        protected abstract Msg tryTake();

        /// GUEST TO HOST. The sandbox sent these bytes on that port.
        protected abstract void put(int port, byte[] wire);

        /// The guest called `open` on a capability, on the port it asked through.
        /// Return the host port id to GRANT, or 0 to refuse -- a refusal is a
        /// normal outcome and surfaces in the program as a catchable error.
        ///
        /// A GRANT NAMES A PORT, which is why this answers an id rather than a
        /// boolean: there is no port until the host says which one.
        protected int open(int port, String capability, byte[] args) { return 0; }

        /// The guest asked the host for a value that is not a port
        /// (`DECISIONS.md#workspace-capabilities` step 7). Return the encoded
        /// value, or null to refuse.
        protected byte[] answer(int port, byte[] wire) { return null; }

        /// The guest closed its end of that port.
        protected void closed(int port) {}
    }

    // --- status ------------------------------------------------------------

    /// `loop`'s answers, and THE EXISTING ABI NUMBERS -- this project treats them
    /// as the ABI itself, so they are not re-enumerated here.
    ///
    /// `NEEDS_HOST` IS THE RESTING STATE, not an error. The control plane is a
    /// green thread parked on the system port, so a healthy idle sandbox reports
    /// it.
    /// `SHELVED` and `WEDGED` added 2026-09-25. 3 was already returned by the
    /// runtime (`snap.rs`'s `STATUS_SHELVED`) and declared by no face; 4 is new,
    /// because the scheduler's deadlock branch answered 0 -- the same number a clean
    /// finish answers -- so a host could not tell a settled sandbox from a wedged
    /// one. WEDGED is reachable only for a sandbox with NO system port: one with a
    /// door answers NEEDS_HOST and lets the host decide it has nothing left to send.
    public static final int DONE = 0, THREW = 1, NEEDS_HOST = 2,
                           SHELVED = 3, WEDGED = 4;

    // --- state -------------------------------------------------------------

    /// Overrides waiting for a boot, and whether one has happened.
    ///
    /// STATIC, because `link` has to be callable with no sandbox in hand -- it
    /// must precede `boot`. One sandbox per class loader, which is the same reason
    /// the generated `flint.Artifact` keeps its sandbox in a static field.
    private static final Map<String, Builtins.Fn> OVERRIDES = new LinkedHashMap<>();
    private static Sandbox current;

    private final Rt rt;
    private final Bridge bridge;

    private Sandbox(Rt rt, Bridge bridge) { this.rt = rt; this.bridge = bridge; }

    // --- 1. boot -----------------------------------------------------------

    /// Load the artifact's image and give the sandbox its door.
    ///
    /// `imageChunks` IS PASSED IN: the bytes are constant data on the generated
    /// `flint.Artifact` class, which calls this. Nothing here opens a resource, so
    /// the program is not a file the artifact contains -- it is data the artifact is
    /// made of, and the CLR carries it the same way.
    ///
    /// IT FAILS ON A MISSING NATIVE, naming it. See below.
    ///
    /// ONE PROGRAM FOR THE SANDBOX'S WHOLE LIFE
    /// (`DECISIONS.md#construe-integration-bar`). Booting twice is refused rather
    /// than replacing the image: a second program in the same sandbox would share
    /// a heap with the first.
    public static Sandbox boot(Bridge system, String[] imageChunks) {
        if (current != null) throw new IllegalStateException(
            "this sandbox is already booted: a sandbox is ONE program for its whole life");
        Rt rt = new Rt(NURSERY, HEAP);
        Sandbox box = new Sandbox(rt, system);
        byte[] image = decode(imageChunks);
        // THE `Loaded` IS KEPT NOW. It carries the image's `flags` word, and this
        // line discarded it -- so `FLAG_PERF` was parsed at `Img.java:140`, stored
        // in a field, and read by nobody. See the `compileArities` call below.
        Img.Loaded loaded = Img.load(rt, image);
        if (loaded == null) throw new IllegalStateException(
            "the artifact's image is not a flint image, or is a version this runtime does not"
            + " speak (this runtime reads version " + Img.VERSION + ", got " + image.length
            + " bytes in " + imageChunks.length + " chunk(s))");

        // NATIVES RESOLVE EXACTLY ONCE, here. `Img.load` has just bound each slot
        // through `Builtins.byName` -- the host's own table, which is what "the
        // host carries the runtime" means on this target -- and the overrides
        // `link` collected go on top before anything runs. That is why `link`
        // refuses after this point: there is no second resolution to join.
        //
        // A NAME NOBODY ANSWERS FOR STAYS NULL rather than failing the load. An
        // image imports every builtin its namespaces MENTION, and a program that
        // never calls the missing one runs fine; the failure names it if reached.
        int n = rt.nativeNames == null ? 0 : rt.nativeNames.length;
        java.util.List<String> missing = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            // `containsKey`, NOT a null check on `get`. An entry whose value is null
            // means the host declared the native ABSENT, and reading that as "no
            // override" would silently fall back to the host's own -- turning a
            // deliberate declaration into a no-op and making the boot failure
            // untestable.
            if (OVERRIDES.containsKey(rt.nativeNames[i])) {
                rt.natives[i] = OVERRIDES.get(rt.nativeNames[i]);
            }
            if (rt.natives[i] == null) missing.add(rt.nativeNames[i]);
        }
        // BOOT FAILS ON A MISSING NATIVE, and this is a deliberate change away from
        // `Img.load`'s null-tolerance. `Img` leaves a slot null on purpose -- an
        // image imports every builtin its namespaces MENTION, so a program that
        // never calls the missing one runs fine -- and the record chose the other
        // trade: the VERSION says what must be linked, so a gap is a version
        // mismatch and is worth refusing now rather than when something reaches it.
        //
        // A version-matched host carries all 88 a trivial program declares and
        // nothing changes. A host that deliberately TRIMMED its builtins now fails
        // to boot a program it could have run; that is the cost, and it is chosen.
        //
        // THE MESSAGE NAMES THEM, capped, because a host whose table is a version
        // behind is missing a handful and a host built wrong is missing dozens --
        // and a failure that says "8 natives are missing" sends someone reading the
        // wrong file.
        if (!missing.isEmpty()) {
            // `current` STAYS UNSET, so a host can `link` what it was missing and
            // try again -- which is what `FourOps` does to prove this path. It is
            // unset rather than cleared: the guard at the top already refused a
            // second boot, so nothing has been assigned yet, and a `current = null`
            // here would read as cleanup that does nothing.
            int shown = Math.min(missing.size(), 8);
            throw new IllegalStateException(
                "this runtime does not carry " + missing.size() + " of the "
                + n + " natives this program declares: "
                + String.join(", ", missing.subList(0, shown))
                + (missing.size() > shown ? ", ... (" + (missing.size() - shown) + " more)" : "")
                + ". The artifact's version says what has to be linked, so this is a"
                + " version mismatch between it and this host -- read"
                + " com.3sln.flint.meta on the artifact's class for the version it"
                + " was built against.");
        }

        // COMPILED ARITIES, when the image asks for them. `:optimize [perf]` sets
        // `FLAG_PERF` in the image (`src/flint/image.cljc`), and until now nothing
        // on this port acted on it: `AotEmit` is 408 lines that ran only from
        // `runtimes/jvm/test/RtAot.java`, so every artifact was interpreted however
        // it was compiled. The producer's own comment says the case was built for
        // this -- "an image for a PORT has the bit and an empty table, which is
        // exactly the case that could not be expressed before".
        //
        // HERE, because natives are resolved above and nothing has executed yet:
        // `RtAot.java` establishes the order as load, compile, then initialisers,
        // and the initialisers do not run until the first `loop`.
        //
        // `false` for `chunkAll`: it makes EVERY instruction a chunk boundary and
        // is "a bisection handle, not a mode" in `AotPlan`'s own words. The
        // reference producer agrees -- `src/flint/aot.cljc`'s four-argument
        // `compile-arity` delegates with `false`, and `src/flint/bundle.cljc` calls
        // that form.
        //
        // AN ARITY THAT CANNOT BE COMPILED STAYS INTERPRETED, so this cannot fail a
        // boot that would otherwise have worked: `AotEmit.compile` answers null for
        // an opcode it does not know and `compileArities` skips it.
        if ((loaded.flags & Img.FLAG_PERF) != 0) {
            rt.compileArities(false);
        }

        Conc.installSystemPort(rt, system.systemPort, Str.of(rt, system.label));
        current = box;
        return box;
    }

    /// The heap the sandbox boots with.
    ///
    /// FIXED, and not read off the artifact. They were metadata for a while:
    /// `nursery`, `heap` and `gas` rode in the properties so `boot` could apply what
    /// the builder chose. Nothing ever set them, and once `prop` was gone there was
    /// no way for a host to act on them either -- so they were metadata carried and
    /// never read, which is exactly the thing this project has a commit about. When
    /// a program needs its own bounds, they become arguments to `boot` or a second
    /// attribute the HOST reads and applies; a constant here is honest until then.
    static final long NURSERY = 1024 * 1024, HEAP = 64L * 1024 * 1024;

    /// The image, out of the string constants that carry it.
    ///
    /// ONE CHARACTER PER BYTE, so this is a cast and a copy. The class file stores
    /// each character in modified UTF-8 -- one byte for `0x01`..`0x7F`, two for
    /// `0x00` and for `0x80`..`0xFF` -- and `charAt` gives the character back, so
    /// nothing here decodes anything. The chunking exists only because a
    /// `CONSTANT_Utf8_info` length is a `u2`.
    private static byte[] decode(String[] chunks) {
        int n = 0;
        for (String s : chunks) n += s.length();
        byte[] out = new byte[n];
        int at = 0;
        for (String s : chunks) {
            for (int i = 0; i < s.length(); i++) out[at++] = (byte) s.charAt(i);
        }
        return out;
    }

    // --- 2. loop -----------------------------------------------------------

    /// Pump. `DONE`, `THREW` or `NEEDS_HOST`.
    ///
    /// DRAIN, DRIVE, DRAIN AGAIN, which the record is explicit about: the
    /// scheduler reports `NEEDS_HOST` while it is still holding undrained events,
    /// so a pump that skipped the trailing drain would report work it was itself
    /// holding and the host would answer a question nobody had asked yet.
    ///
    /// The inbound pull comes first, so whatever the host queued is in the
    /// sandbox's inboxes before the scheduler looks for something runnable.
    ///
    /// IT DOES NOT RETURN THE ANSWER. A program's answer comes back as a message
    /// on a bridge, because a bridge is the only door
    /// (`DECISIONS.md#bridges-are-the-only-door`) -- and a `loop` that also
    /// answered would only work for programs with a distinguished entry point.
    /// Those do not exist: a caller names the function it wants.
    ///
    /// THE LOOP CONDITION IS PROGRESS, not a spin count.
    /// `sdks/esm/src/guest.js` loops while the code is 2 and gives up at a
    /// million; here the question is whether this turn moved anything, which is
    /// the actual question and needs no ceiling.
    public int loop() {
        for (;;) {
            boolean took = pull();
            boolean before = drain();
            Conc.drive(rt);
            boolean after = drain();
            if (rt.status != NEEDS_HOST) {
                // THE BOUNDARY LOOKS AT `thrown` TO DECIDE STATUS, and
                // `Mainanswer.mainAnswer` is what puts it there: `TH_RESULT`
                // holds the error for a failed thread and the value for a
                // finished one, so the status cannot be read off the answer.
                return Val.isNil(rt.thrown) ? DONE : THREW;
            }
            if (!took && !before && !after) return NEEDS_HOST;
        }
    }

    /// Everything the host has queued, into the sandbox. True if anything moved.
    private boolean pull() {
        boolean any = false;
        for (Msg m = bridge.tryTake(); m != null; m = bridge.tryTake()) {
            Conc.hostDeliver(rt, m.port(), m.bytes());
            any = true;
        }
        return any;
    }

    /// Everything the guest said, to the bridge. True if anything moved.
    ///
    /// The five-`u32` records are unpacked HERE rather than by the host. A wasm
    /// host has to do this itself, over linear memory, at `flint_events_ptr`.
    private boolean drain() {
        Conc.Events evs = Conc.drainEvents(rt);
        if (evs.count() == 0) return false;
        byte[] b = evs.bytes();
        for (int i = 0; i < evs.count(); i++) {
            int rec = i * 20;
            int kind = (int) u32(b, rec), a = (int) u32(b, rec + 4), bb = (int) u32(b, rec + 8);
            int off = (int) u32(b, rec + 12), len = (int) u32(b, rec + 16);
            byte[] payload = new byte[len];
            System.arraycopy(b, off, payload, 0, len);
            switch (kind) {
                case Conc.EV_MESSAGE -> bridge.put(a, payload);
                case Conc.EV_CLOSED -> bridge.closed(a);
                // `bb` is the SYSTEM port the request came out on, not a port
                // made for it: there is no port until the host grants one.
                case Conc.EV_OPEN -> {
                    int granted = bridge.open(bb, capabilityName(payload), payload);
                    if (granted == 0 || !Conc.hostGrant(rt, a, granted)) {
                        Conc.hostContinue(rt, a, false);
                    }
                }
                case Conc.EV_REQUEST -> {
                    byte[] answer = bridge.answer(bb, payload);
                    if (answer == null || !Conc.hostAnswer(rt, a, answer)) {
                        Conc.hostContinue(rt, a, false);
                    }
                }
                // EV_RETAIN / EV_RELEASE are the host's reference count on its own
                // ports. A host sharing a port between sandboxes needs them; this
                // sandbox holds every port it was handed for its whole life, so a
                // count of holders has nothing to say to it.
                default -> { }
            }
        }
        return true;
    }

    // --- 3. link -----------------------------------------------------------

    /// Override the native called `name`.
    ///
    /// AN OVERRIDE MECHANISM AND NOT A WIRING ONE, and the difference is the whole
    /// design. A trivial program's image declares 88 natives -- `clojure.core`'s
    /// reach, not the program's -- against 223 the runtime carries, and a host
    /// makes ZERO `link` calls in the normal case because `Img.load` resolves them
    /// all against `Builtins.byName`. A bulk or resolver API that required all 88
    /// would be actively wrong: a missing native is left NULL rather than refused,
    /// so demanding every one of them would reject programs that run.
    ///
    /// IT MUST PRECEDE `boot` and refuses afterwards, because natives resolve
    /// exactly once when the image loads. Refusing is the point: an override
    /// registered after boot would silently not apply, and a host would be left
    /// with a hook it believed in.
    public static void link(String name, Builtins.Fn fn) {
        if (current != null) throw new IllegalStateException(
            "link(" + name + ") after boot: natives resolve exactly once when the image loads,"
            + " so an override registered now would never apply. Call link before boot.");
        // A NULL `fn` DECLARES THE NATIVE ABSENT, and does not forget the override.
        // That is the only sense it can have now that `boot` fails on a missing
        // native: a host saying "I do not carry this" is saying something, and it is
        // the only way anything can exercise that failure through this surface.
        // Reverting to the host's own is `link(name, Builtins.byName(name))`.
        OVERRIDES.put(name, fn);
    }

    // --- there is no fourth ---
    //
    // `prop(name, buf)` lived here. It answered a metadata property, and the
    // metadata is now a class attribute the CONTAINER carries
    // (`DECISIONS.md#four-operations`) -- so `readProps`, `propLong`, the
    // `name<TAB>value` flattening and the NO-LINKAGE-REPORT tally went with it.
    //
    // What replaced the tally is `boot` failing above. An earlier face grew
    // `prop("natives")` and `prop("natives-unresolved")` so a host with a trimmed
    // runtime could discover a gap before a program reached it; refusing to boot
    // tells it the same thing at the same moment and needs no operation.

    /// The capability an `open` asked for, out of the encoded argument vector.
    ///
    /// NOT A DECODER. The payload is `[name & args]` as one encoded value, and
    /// reading the name properly means the wire codec -- which belongs in an SDK
    /// beside `sdks/esm/src/codec.js`, not in the runtime. This finds the first
    /// string so a host's `open` has something to switch on, and the whole payload
    /// goes with it so a host with a real decoder can ignore this.
    private static String capabilityName(byte[] payload) {
        for (int i = 0; i + 5 <= payload.length; i++) {
            if ((payload[i] & 0xff) != 5) continue;          // Codec.K_STRING
            long n = u32(payload, i + 1);
            if (n <= 0 || i + 5 + n > payload.length) continue;
            return new String(payload, i + 5, (int) n, StandardCharsets.UTF_8);
        }
        return "";
    }

    private static long u32(byte[] b, int at) {
        return (b[at] & 0xffL) | ((b[at + 1] & 0xffL) << 8)
             | ((b[at + 2] & 0xffL) << 16) | ((b[at + 3] & 0xffL) << 24);
    }
}
